File: examples/api-example.py
#!/usr/bin/env python3
"""
Etu Server API Example
Demonstrates how to interact with the Etu Server REST API using Python.
"""

import requests
import json
from typing import Optional, Dict, List, Any

# Configuration
API_KEY = "etu_your_api_key_here"
BASE_URL = "http://localhost:3000/api/v1"

class EtuClient:
    """Simple client for the Etu Server API."""
    
    def __init__(self, api_key: str, base_url: str = BASE_URL):
        self.api_key = api_key
        self.base_url = base_url
        self.headers = {
            "Authorization": api_key,
            "Content-Type": "application/json"
        }
    
    def list_notes(
        self,
        search: Optional[str] = None,
        tags: Optional[List[str]] = None,
        limit: int = 50,
        offset: int = 0
    ) -> Dict[str, Any]:
        """List notes with optional filtering."""
        params = {"limit": limit, "offset": offset}
        if search:
            params["search"] = search
        if tags:
            params["tags"] = ",".join(tags)
        
        response = requests.get(
            f"{self.base_url}/notes",
            headers=self.headers,
            params=params
        )
        response.raise_for_status()
        return response.json()
    
    def create_note(self, content: str, tags: Optional[List[str]] = None) -> Dict[str, Any]:
        """Create a new note."""
        data = {"content": content}
        if tags:
            data["tags"] = tags
        
        response = requests.post(
            f"{self.base_url}/notes",
            headers=self.headers,
            json=data
        )
        response.raise_for_status()
        return response.json()
    
    def get_note(self, note_id: str) -> Dict[str, Any]:
        """Get a specific note by ID."""
        response = requests.get(
            f"{self.base_url}/notes/{note_id}",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def update_note(
        self,
        note_id: str,
        content: Optional[str] = None,
        tags: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """Update a note."""
        data = {}
        if content is not None:
            data["content"] = content
        if tags is not None:
            data["tags"] = tags
        
        response = requests.put(
            f"{self.base_url}/notes/{note_id}",
            headers=self.headers,
            json=data
        )
        response.raise_for_status()
        return response.json()
    
    def delete_note(self, note_id: str) -> Dict[str, Any]:
        """Delete a note."""
        response = requests.delete(
            f"{self.base_url}/notes/{note_id}",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def list_tags(self) -> Dict[str, Any]:
        """List all tags."""
        response = requests.get(
            f"{self.base_url}/tags",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()


def main():
    """Example usage of the Etu API client."""
    print("Etu Server API Example")
    print("=" * 50)
    
    # Initialize client
    client = EtuClient(API_KEY)
    
    try:
        # Example 1: Create a note
        print("\n1. Creating a new note...")
        note = client.create_note(
            content="# Meeting Notes\n\n- Discussed project timeline\n- Reviewed requirements",
            tags=["work", "meeting"]
        )
        print(f"   Created note: {note['id']}")
        note_id = note['id']
        
        # Example 2: Get the note
        print("\n2. Retrieving the note...")
        retrieved = client.get_note(note_id)
        print(f"   Content: {retrieved['content'][:50]}...")
        
        # Example 3: Update the note
        print("\n3. Updating the note...")
        updated = client.update_note(
            note_id,
            content="# Meeting Notes (Updated)\n\n- Discussed project timeline\n- Reviewed requirements\n- Assigned action items"
        )
        print(f"   Updated at: {updated['updatedAt']}")
        
        # Example 4: List all notes
        print("\n4. Listing all notes...")
        notes = client.list_notes(limit=5)
        print(f"   Total notes: {notes['total']}")
        print(f"   Fetched: {len(notes['notes'])} notes")
        
        # Example 5: Search notes
        print("\n5. Searching for notes...")
        results = client.list_notes(search="meeting")
        print(f"   Found {len(results['notes'])} notes matching 'meeting'")
        
        # Example 6: Filter by tags
        print("\n6. Filtering notes by tags...")
        tagged = client.list_notes(tags=["work"])
        print(f"   Found {len(tagged['notes'])} notes with 'work' tag")
        
        # Example 7: List all tags
        print("\n7. Listing all tags...")
        tags = client.list_tags()
        print(f"   Total tags: {len(tags['tags'])}")
        for tag in tags['tags'][:5]:
            print(f"   - {tag['name']}: {tag['count']} notes")
        
        # Example 8: Delete the note
        print(f"\n8. Deleting note {note_id}...")
        result = client.delete_note(note_id)
        print(f"   Success: {result['success']}")
        
        print("\n" + "=" * 50)
        print("All examples completed successfully!")
        
    except requests.exceptions.HTTPError as e:
        print(f"\nError: {e}")
        print(f"Response: {e.response.text}")
    except Exception as e:
        print(f"\nUnexpected error: {e}")


if __name__ == "__main__":
    main()

File: examples/api-example.js
#!/usr/bin/env node
/**
 * Etu Server API Example
 * Demonstrates how to interact with the Etu Server REST API using JavaScript/Node.js.
 */

// Configuration
const API_KEY = "etu_your_api_key_here";
const BASE_URL = "http://localhost:3000/api/v1";

class EtuClient {
  constructor(apiKey, baseUrl = BASE_URL) {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
    this.headers = {
      "Authorization": apiKey,
      "Content-Type": "application/json",
    };
  }

  async listNotes({ search, tags, limit = 50, offset = 0 } = {}) {
    const params = new URLSearchParams({ limit, offset });
    if (search) params.append("search", search);
    if (tags) params.append("tags", tags.join(","));

    const response = await fetch(`${this.baseUrl}/notes?${params}`, {
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    return response.json();
  }

  async createNote(content, tags = []) {
    const response = await fetch(`${this.baseUrl}/notes`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify({ content, tags }),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    return response.json();
  }

  async getNote(noteId) {
    const response = await fetch(`${this.baseUrl}/notes/${noteId}`, {
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    return response.json();
  }

  async updateNote(noteId, { content, tags } = {}) {
    const data = {};
    if (content !== undefined) data.content = content;
    if (tags !== undefined) data.tags = tags;

    const response = await fetch(`${this.baseUrl}/notes/${noteId}`, {
      method: "PUT",
      headers: this.headers,
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    return response.json();
  }

  async deleteNote(noteId) {
    const response = await fetch(`${this.baseUrl}/notes/${noteId}`, {
      method: "DELETE",
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    return response.json();
  }

  async listTags() {
    const response = await fetch(`${this.baseUrl}/tags`, {
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    return response.json();
  }
}

async function main() {
  console.log("Etu Server API Example");
  console.log("=".repeat(50));

  // Initialize client
  const client = new EtuClient(API_KEY);

  try {
    // Example 1: Create a note
    console.log("\n1. Creating a new note...");
    const note = await client.createNote(
      "# Daily Standup\n\n- Completed feature X\n- Working on feature Y\n- Blocked on issue Z",
      ["work", "standup"]
    );
    console.log(`   Created note: ${note.id}`);
    const noteId = note.id;

    // Example 2: Get the note
    console.log("\n2. Retrieving the note...");
    const retrieved = await client.getNote(noteId);
    console.log(`   Content: ${retrieved.content.substring(0, 50)}...`);
    console.log(`   Tags: ${retrieved.tags.join(", ")}`);

    // Example 3: Update the note
    console.log("\n3. Updating the note...");
    const updated = await client.updateNote(noteId, {
      content:
        "# Daily Standup (Updated)\n\n- Completed feature X ✅\n- Working on feature Y\n- Issue Z resolved",
    });
    console.log(`   Updated at: ${updated.updatedAt}`);

    // Example 4: List all notes
    console.log("\n4. Listing all notes...");
    const notes = await client.listNotes({ limit: 5 });
    console.log(`   Total notes: ${notes.total}`);
    console.log(`   Fetched: ${notes.notes.length} notes`);

    // Example 5: Search notes
    console.log("\n5. Searching for notes...");
    const results = await client.listNotes({ search: "standup" });
    console.log(`   Found ${results.notes.length} notes matching 'standup'`);

    // Example 6: Filter by tags
    console.log("\n6. Filtering notes by tags...");
    const tagged = await client.listNotes({ tags: ["work"] });
    console.log(`   Found ${tagged.notes.length} notes with 'work' tag`);

    // Example 7: List all tags
    console.log("\n7. Listing all tags...");
    const tags = await client.listTags();
    console.log(`   Total tags: ${tags.tags.length}`);
    tags.tags.slice(0, 5).forEach((tag) => {
      console.log(`   - ${tag.name}: ${tag.count} notes`);
    });

    // Example 8: Delete the note
    console.log(`\n8. Deleting note ${noteId}...`);
    const result = await client.deleteNote(noteId);
    console.log(`   Success: ${result.success}`);

    console.log("\n" + "=".repeat(50));
    console.log("All examples completed successfully!");
  } catch (error) {
    console.error("\nError:", error.message);
  }
}

// Run the examples
main();
